hudi-common/src/main/java/org/apache/hudi/common/model/HoodieCommitMetadata.java

```diff
@@ -175,7 +175,9 @@ public long fetchTotalInsertRecordsWritten() {
     long totalInsertRecordsWritten = 0;
     for (List<HoodieWriteStat> stats : partitionToWriteStats.values()) {
       for (HoodieWriteStat stat : stats) {
-        if (stat.getPrevCommit() != null && stat.getPrevCommit().equalsIgnoreCase("null")) {
```

Comment:
For the first commit, the prevCommmit value is actually "null" which is the reason why this check is here. 
Can you inspect and see if stat.getPrevCommit() returns null in some cases ? Then, the above condition will not be satisfied.

Reply:
@n3nash
I think the processing logic of this method is the same as that of `fetchTotalFilesUpdated`
```
  public long fetchTotalUpdateRecordsWritten() {
    long totalUpdateRecordsWritten = 0;
    for (List<HoodieWriteStat> stats : partitionToWriteStats.values()) {
      for (HoodieWriteStat stat : stats) {
        totalUpdateRecordsWritten += stat.getNumUpdateWrites();
      }
    }
    return totalUpdateRecordsWritten;
  }
```
Let me give you an example in my test. I create a hudi data with 3 rows first, and then insert a row, at last insert a row and delete a row. I print all method return number and test the number to check.
```java
for (HoodieInstant commit : commits) {
    HoodieCommitMetadata commitMetadata = HoodieCommitMetadata.fromBytes(timeline.getInstantDetails(commit).get(),
            HoodieCommitMetadata.class);

    JSONObject data = new JSONObject();
    data.put("timestamp", commit.getTimestamp());
    data.put("fetchTotalBytesWritten", commitMetadata.fetchTotalBytesWritten());
    data.put("fetchTotalFilesInsert", commitMetadata.fetchTotalFilesInsert());
    data.put("fetchTotalFilesUpdated", commitMetadata.fetchTotalFilesUpdated());
    data.put("fetchTotalPartitionsWritten", commitMetadata.fetchTotalPartitionsWritten());
    data.put("fetchTotalRecordsWritten", commitMetadata.fetchTotalRecordsWritten());
    data.put("fetchTotalUpdateRecordsWritten", commitMetadata.fetchTotalUpdateRecordsWritten());
    data.put("fetchTotalInsertRecordsWritten", commitMetadata.fetchTotalInsertRecordsWritten());
    data.put("fetchTotalWriteErrors", commitMetadata.fetchTotalWriteErrors());

    long totalDeleteRecordsWritten= 0;
    for (List<HoodieWriteStat> stats : commitMetadata.getPartitionToWriteStats().values()) {
        for (HoodieWriteStat stat : stats) {
            if (stat.getPrevCommit() != null) {
                totalDeleteRecordsWritten += stat.getNumDeletes();
            }
        }
    }
    data.put("fetchTotalDeleteRecordsWritten", totalDeleteRecordsWritten);

    datas.add(data);
}
log.debug("datas = {}", datas);
```
If I user old if `(stat.getPrevCommit() != null && stat.getPrevCommit().equalsIgnoreCase("null"))` it will print:
```json
[
	{
		"fetchTotalBytesWritten": 435258,
		"fetchTotalDeleteRecordsWritten": 1,
		"fetchTotalFilesInsert": 0,
		"fetchTotalFilesUpdated": 1,
		"fetchTotalInsertRecordsWritten": 0,
		"fetchTotalPartitionsWritten": 1,
		"fetchTotalRecordsWritten": 4,
		"fetchTotalUpdateRecordsWritten": 0,
		"fetchTotalWriteErrors": 0,
		"timestamp": "20190923172508"
	},
	{
		"fetchTotalBytesWritten": 435227,
		"fetchTotalDeleteRecordsWritten": 0,
		"fetchTotalFilesInsert": 0,
		"fetchTotalFilesUpdated": 1,
		"fetchTotalInsertRecordsWritten": 0,
		"fetchTotalPartitionsWritten": 1,
		"fetchTotalRecordsWritten": 4,
		"fetchTotalUpdateRecordsWritten": 0,
		"fetchTotalWriteErrors": 0,
		"timestamp": "20190923160439"
	},
	{
		"fetchTotalBytesWritten": 435478,
		"fetchTotalDeleteRecordsWritten": 0,
		"fetchTotalFilesInsert": 1,
		"fetchTotalFilesUpdated": 0,
		"fetchTotalInsertRecordsWritten": 3,
		"fetchTotalPartitionsWritten": 1,
		"fetchTotalRecordsWritten": 3,
		"fetchTotalUpdateRecordsWritten": 0,
		"fetchTotalWriteErrors": 0,
		"timestamp": "20190923160340"
	}
]
```
If I user modified if `(stat.getPrevCommit() != null)` it will print:
```json
[
	{
		"fetchTotalBytesWritten": 435258,
		"fetchTotalDeleteRecordsWritten": 1,
		"fetchTotalFilesInsert": 0,
		"fetchTotalFilesUpdated": 1,
		"fetchTotalInsertRecordsWritten": 1,
		"fetchTotalPartitionsWritten": 1,
		"fetchTotalRecordsWritten": 4,
		"fetchTotalUpdateRecordsWritten": 0,
		"fetchTotalWriteErrors": 0,
		"timestamp": "20190923172508"
	},
	{
		"fetchTotalBytesWritten": 435227,
		"fetchTotalDeleteRecordsWritten": 0,
		"fetchTotalFilesInsert": 0,
		"fetchTotalFilesUpdated": 1,
		"fetchTotalInsertRecordsWritten": 1,
		"fetchTotalPartitionsWritten": 1,
		"fetchTotalRecordsWritten": 4,
		"fetchTotalUpdateRecordsWritten": 0,
		"fetchTotalWriteErrors": 0,
		"timestamp": "20190923160439"
	},
	{
		"fetchTotalBytesWritten": 435478,
		"fetchTotalDeleteRecordsWritten": 0,
		"fetchTotalFilesInsert": 1,
		"fetchTotalFilesUpdated": 0,
		"fetchTotalInsertRecordsWritten": 3,
		"fetchTotalPartitionsWritten": 1,
		"fetchTotalRecordsWritten": 3,
		"fetchTotalUpdateRecordsWritten": 0,
		"fetchTotalWriteErrors": 0,
		"timestamp": "20190923160340"
	}
]
```
I guess the person who originally wrote this method copied the `fetchTotalFilesInsert` method at that time, but he forgot that the processing logic of adding files is different from that adding data in files.

Reply:
@n3nash
sorry I may misunderstand your mind, I'll take a look at what I've tested later.

Reply:
@n3nash 
It may be in https://github.com/apache/incubator-hudi/blob/master/hudi-client/src/main/java/org/apache/hudi/io/HoodieCreateHandle.java#L165
```java
      HoodieWriteStat stat = new HoodieWriteStat();
      stat.setPartitionPath(writeStatus.getPartitionPath());
      stat.setNumWrites(recordsWritten);
      stat.setNumDeletes(recordsDeleted);
      stat.setNumInserts(insertRecordsWritten);
      stat.setPrevCommit(HoodieWriteStat.NULL_COMMIT);
      stat.setFileId(writeStatus.getFileId());
      stat.setPath(new Path(config.getBasePath()), path);
      long fileSizeInBytes = FSUtils.getFileSize(fs, path);
      stat.setTotalWriteBytes(fileSizeInBytes);
      stat.setFileSizeInBytes(fileSizeInBytes);
      stat.setTotalWriteErrors(writeStatus.getTotalErrorRecords());
      RuntimeStats runtimeStats = new RuntimeStats();
      runtimeStats.setTotalCreateTime(timer.endTimer());
      stat.setRuntimeStats(runtimeStats);
      writeStatus.setStat(stat);
```
In `org.apache.hudi.common.model.HoodieWriteStat` it is a "null", in other cases prevCommit will be set a real commit time.
```java
public static final String NULL_COMMIT = "null";
```
So that `fetchTotalFilesInsert` can recognize first commit file and pass the condition.

Reply:
@cdmikechen I think the problem is here : The FileSizing of Hoodie takes care of padding extra records to the same file till the file grows to a certain size. This logic lies in HoodieMergeHandle.java. Since the merge handle will always have a prevCommit, for the records that are inserted as part of the HoodieMergeHandle (records padded to an existing file), those will not pass the condition of stat.getPrevCommit().equalsIgnoreCase("null"). So, I think removing the check `stat.getPrevCommit().equalsIgnoreCase("null")` and then adding the num-inserts should be the right thing to do. Let me know if this makes sense to you.

Reply:
@n3nash  Yes, that's almost what it means!